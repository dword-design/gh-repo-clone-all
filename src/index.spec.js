import { endent, join, property, pullAll, split } from '@dword-design/functions'
import proxyquire from '@dword-design/proxyquire'
import tester from '@dword-design/tester'
import execa from 'execa'
import globby from 'globby'
import withLocalTmpDir from 'with-local-tmp-dir'

const pathDelimiter = process.platform === 'win32' ? ';' : ':'

const getModifiedPath = () =>
  process.env.PATH
  |> split(pathDelimiter)
  |> pullAll(['/usr/local/bin', '/usr/bin'])
  |> join(pathDelimiter)

export default tester(
  {
    async branch() {
      const self = proxyquire('.', {
        './gh-repo-list': () => endent`
        dword-design/gh-repo-clone-all-test1
        dword-design/gh-repo-clone-all-test2
      `,
      })
      await self({ branch: 'foo' })
      expect(await globby('*', { onlyFiles: false })).toEqual([
        'gh-repo-clone-all-test1',
        'gh-repo-clone-all-test2',
      ])
      expect(
        execa.command('git status', {
          all: true,
          cwd: 'gh-repo-clone-all-test1',
        })
          |> await
          |> property('all')
      ).toMatchSnapshot(this)
      expect(
        (await execa.command('git status', {
          all: true,
          cwd: 'gh-repo-clone-all-test2',
        }))
          |> await
          |> property('all')
      ).toMatchSnapshot(this)
    },
    cwd: async () => {
      const self = proxyquire('.', {
        './gh-repo-list': () => endent`
        dword-design/gh-repo-clone-all-test1
        dword-design/gh-repo-clone-all-test2
      `,
      })
      expect(await self()).toEqual([])
      expect(await globby('*', { onlyFiles: false })).toEqual([
        'gh-repo-clone-all-test1',
        'gh-repo-clone-all-test2',
      ])
      expect(
        await globby('*', {
          cwd: 'gh-repo-clone-all-test1',
          dot: true,
          onlyFiles: false,
        })
      ).toEqual(['.git'])
      expect(
        await globby('*', {
          cwd: 'gh-repo-clone-all-test2',
          dot: true,
          onlyFiles: false,
        })
      ).toEqual(['.git'])
    },
    'gh missing': async () => {
      const self = proxyquire('.', {})
      process.env = { ...process.env, PATH: getModifiedPath() }
      await expect(self()).rejects.toThrow(
        'It seems like GitHub CLI is not installed on your machine. Install it at https://cli.github.com/manual.'
      )
    },
    'non-existing branch': async function () {
      const self = proxyquire('.', {
        './gh-repo-list': () => endent`
        dword-design/gh-repo-clone-all-test1
        dword-design/gh-repo-clone-all-test2
      `,
      })
      expect(await self({ branch: 'bar' })).toMatchSnapshot(this)
      expect(await globby('*', { onlyFiles: false })).toEqual([
        'gh-repo-clone-all-test2',
      ])
      expect(
        (await execa.command('git status', {
          all: true,
          cwd: 'gh-repo-clone-all-test2',
        }))
          |> await
          |> property('all')
      ).toMatchSnapshot(this)
    },
    'repository not found': async function () {
      const self = proxyquire('.', {
        './gh-repo-list': () => 'foo',
      })
      expect(await self()).toMatchSnapshot(this)
    },
  },
  [
    {
      afterEach() {
        process.env = this.previousEnv
      },
      beforeEach() {
        this.previousEnv = process.env
      },
    },
    {
      transform: test =>
        function () {
          return withLocalTmpDir(() => test.call(this))
        },
    },
  ]
)
