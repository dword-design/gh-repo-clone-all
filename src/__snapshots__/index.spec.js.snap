// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index branch 1`] = `
"On branch foo
Your branch is up to date with 'origin/foo'.

nothing to commit, working tree clean"
`;

exports[`index branch 2`] = `
"On branch foo
Your branch is up to date with 'origin/foo'.

nothing to commit, working tree clean"
`;

exports[`index non-existing branch 1`] = `
Array [
  "Command failed with exit code 128: git clone https://github.com/dword-design/gh-repo-clone-all-test1.git --branch bar
Cloning into 'gh-repo-clone-all-test1'...
fatal: Remote branch bar not found in upstream origin",
]
`;

exports[`index non-existing branch 2`] = `
"On branch bar
Your branch is up to date with 'origin/bar'.

nothing to commit, working tree clean"
`;

exports[`index repository not found 1`] = `
Array [
  "Command failed with exit code 128: git clone https://github.com/foo.git
Cloning into 'foo'...
remote: Not Found
fatal: repository 'https://github.com/foo.git/' not found",
]
`;
